### ISSUE TITLE
Make backends more resilient against keyword and Prelude clashes

### ISSUE THREAD

**Issue by andreasabel** [BNFC/bnfc#278]
OP: Haskell: Printer: variable name "pattern" might clash with PatternSynonyms language extension.
```
Prg. Program ::= Pattern;
PVar. Pattern ::= Ident;
```
creates Haskell printer containing:
```haskell
instance Print Test.Abs.Program where
  prt i e = case e of
    Test.Abs.Prg pattern -> prPrec i 0 (concatD [prt 0 pattern])
```
However, `pattern` is a Haskell keyword when using `{-# LANGUAGE PatternSynonyms #-}`.

When trying to get a test case through that has one category for each keyword of each backend target, I notice a number of related issues:
- [X] Haskell abstract syntax: avoid Prelude clashes
- [X] Haskell template: import Abs module qualified to avoid clashes
- [X] Haskell parser: import Abs qualified
- [X] Agda AST: import Abs qualified in COMPILE pragmas
- [X] Haskell/GADT: same as Haskell
- [X] C: clash of user-defined category `Int` with token type `int` in parser
- [X] C++: same as C
- [X] Ocaml: refer to abstract syntax qualified in print/show
- [X] Java: qualified absyn, names need to be unique modulo case (case insensitive file systems!)
- [X] Java/ANTLR: labels for alternatives may clash with rule names

**Comment by andreasabel** [comment]
Ocamlyacc can currently only handle 127 entry points correctly, see https://github.com/ocaml/ocaml/issues/9207, less than the test case produces.

**Comment by andreasabel** [comment]
Menhir also crashes on this testcase, see https://gitlab.inria.fr/fpottier/menhir/issues/31.

**Comment by andreasabel** [comment]
Both ocamlyacc (shipped with ocaml 4.11.1) and menhir (20200624, my birthday version) are up to the task now.
